indy_node/test/upgrade/test_broken_connection_control_tool.py

```diff
@@ -0,0 +1,159 @@
+import asyncio
+from functools import partial
+
+import dateutil
+import dateutil.tz
+import pytest
+
+from datetime import datetime, timedelta
+
+from copy import deepcopy
+
+from indy_common.constants import START
+
+from indy_node.test.upgrade.conftest import patch_packet_mgr_output, EXT_PKT_NAME, EXT_PKT_VERSION
+
+from indy_common.config_helper import NodeConfigHelper
+
+from indy_node.server.upgrader import Upgrader, UpgradeMessage
+from indy_node.test.helper import TEST_INITIAL_NODE_VERSION
+from indy_node.test.upgrade.helper import sdk_ensure_upgrade_sent, bumpedVersion
+from plenum.test.helper import randomText
+
+from plenum.test.test_node import ensureElectionsDone, checkNodesConnected
+from stp_core.common.log import getlogger
+
+delta = 2
+logger = getlogger()
+
+
+@pytest.fixture(scope="module")
+def tconf(tconf):
+    old_delta = tconf.MinSepBetweenNodeUpgrades
+    tconf.MinSepBetweenNodeUpgrades = delta
+    yield tconf
+    tconf.MinSepBetweenNodeUpgrades = old_delta
+
+
+@pytest.fixture(scope='function')
+def validUpgrade(nodeIds, tconf, monkeypatch):
+    schedule = {}
+    unow = datetime.utcnow().replace(tzinfo=dateutil.tz.tzutc())
+    startAt = unow + timedelta(seconds=100)
+    acceptableDiff = tconf.MinSepBetweenNodeUpgrades + 1
+    for i in nodeIds:
+        schedule[i] = datetime.isoformat(startAt)
+        startAt = startAt + timedelta(seconds=acceptableDiff + 3)
+
+    patch_packet_mgr_output(monkeypatch, EXT_PKT_NAME, EXT_PKT_VERSION)
+
+    return dict(name='upgrade-{}'.format(randomText(3)), version=bumpedVersion(EXT_PKT_VERSION),
+                action=START, schedule=schedule, timeout=1, package=EXT_PKT_NAME,
+                sha256='db34a72a90d026dae49c3b3f0436c8d3963476c77468ad955845a1ccf7b03f55')
+
+
+@pytest.fixture(scope='function')
+def skip_functions():
+    # Do this to prevent exceptions because of node_control_tool absence
+    old_get_version = deepcopy(Upgrader.getVersion)
+    old_update = deepcopy(Upgrader._update_action_log_for_started_action)
+    Upgrader.getVersion = lambda self, pkg: '1.6.0'
+    Upgrader._update_action_log_for_started_action = lambda self: 1
+    yield
+    Upgrader.getVersion = staticmethod(old_get_version)
+    Upgrader._update_action_log_for_started_action = old_update
+
+
+async def _sendUpgradeRequest(self, when, version, upgrade_id, failTimeout, pkg_name):
```

Comment:
Why do we need so many overrides and patches? Can't we just have a more simple test similar to other tests in `upgrade` folder? For example, `test_upgrade_does_not_get_into_loop`?

Reply:
fixed